src/client/dfuse/inval.c

```diff
@@ -0,0 +1,423 @@
+ * consumption tends to zero.
+ *
+ * For kernel based filesystems there is a benefit to keeping all this data in memory as it can
+ * simply be re-validated before use however with fuse + DAOS then re-validate is the same cost
+ * as lookup so there is no benefit in keeping this data around.
+ *
+ * Maintain a number of lists for inode timeouts, for each timeout value keep a list of inodes
+ * that are using that value, when a inode is refreshed by the kernel then move the inode to the end
+ * of the correct list.
+ *
+ * Separately have a thread which periodically will walk each list starting at the front and
+ * invalidate any entries where the timeout has expired.
+ *
+ * In this way the lists are never traversed, on access a entry is removed from where it is and
+ * appended to the end, and the timeout starts at the front of the list and traverses only as far
+ * as it needs to until the front entry is to be kept.
+ *
+ * As lookups will not be repeated by the kernel until after timeout has expired allow some leeway
+ * before eviction to allow re-validation of in-use datasets without triggering entire tree
+ * invalidations through the kernel.  Directories get five seconds, anything else two.  Ideally
+ * directories would be invalidated first as this would result in less dfuse->kernel calls as once
+ * the kernel invalidates directories then it invalidates the whole tree below that, however there
+ * are also use-cases where there are significiant numbers of files per directory where the
+ * directory is in active use but individual files are not.
+ *
+ * Future work might be to speculatively perform lookups close to the end of the timeout period,
+ * then if a entry was in frequent use it's lookup could be performed from memory, effectively
+ * moving the re-validation cost off the critical path.  This code currently only handles dentries
+ * but could also separately track attributes (inodes) and file contents as well.
+ *
+ * Additional changes to consider in the future could include:
+ *  Better handing of eviction timeouts, "max(time * 1.1, 10)" would be better than a flat +2/+5
+ *  Use arrays rather than lists for the buckets for faster iteration.
+ *  Reference counting the timeout buckets.
+ *
+ * Locking: The ival_lock is contended, it is accessed several places, however none do any more
+ * than list management.  As inodes might be removed from one list and re-inserted into another
+ * there is a pre subsystem lock rather than per list locks.
+ *  ie_close() which is called from forget and some failure paths in readdir()
+ *  lookup() to move entries to the end of this list.
+ *  de_run() to pull items from the front of the list.
+ *
+ * Wakeup: The invalidation thread is woken up when:
+ *  dfuse is exiting.
+ *  something is added to an empty list.
+ *  after a timeout.
+ * Timeouts are chosen based on the entries still on any list, dfuse will sleep as long as it can
+ * but at least 2 seconds and at most 60.
```

Comment:
Maybe I missed it, but I don't see the "at most 60"

Reply:
Line 124.  It starts at 59, then decreases based on any inode with a lower timeout, then rounds up.